### PR TITLE
feat(mlcache) custom LRU layers

### PR DESCRIPTION
**Pull request by thibaultcha**
We can now instruct mlcache to cache a given value in user-specified
LRU. One can now namespace entities, and store entities of various
sizes, hit/miss frequency in different Lua-land LRUs. One can also use
the `pureffi` LRU implementation if desired.

```lua
local resty_lru = require "resty.lrucache.pureffi"
local my_lru = resty_lru.new(10e3)

local cache = mlcache.new()

cache:add_lru("my_pureffi_lru", my_lru)

local value, err = cache:get("my_key", { lru = "my_pureffi_lru" }, cb)
if err then
    error(err)
end

-- value is stored in the custom lru
print(my_lru:get("my_key")) -- "123"
```

### CODE REVIEW

**Comment by thibaultcha** [comment]
@tieske @p0pr0ck5 Do you like this approach? Feedback welcome!

File: lib/resty/mlcache.lua [hunk 1]
```diff
@@ -170,25 +172,48 @@ function _M.new(shm, opts)
...
+        return error("lru appears not to be a lua-resty-lrucache instance")
+    end
+
+    if self.lrus[name] then
+        return error("an lru named '" .. name .. "' has already been added")
+    end
+
+    self.lrus[name] = lru
+
+    insert(self.lrus, lru)
```
Comment (p0pr0ck5) [inline lib/resty/mlcache.lua]: style: insert() here, but elsewhere #self.lrus. why not `self.lrus[#self.lrus] = lru` here?

related: do we care enough about micro-optimizations to store an lrus table index as part of self, rather than than recomputing `#lrus`

File: lib/resty/mlcache.lua [hunk 2]
```diff
@@ -286,15 +311,16 @@ function _M:get(key, opts, cb, ...)
 
     -- opts validation
 
+    local lru
     local ttl
     local neg_ttl
 
-    if opts then
+    if opts ~= nil then
```
Comment (p0pr0ck5) [inline lib/resty/mlcache.lua]: unless theres an explicit reason for this change, it seems this change in type checking belongs in a separate commit.

File: lib/resty/mlcache.lua (hunk 1 above)
Comment (thibaultcha) [inline lib/resty/mlcache.lua]: > style: insert() here, but elsewhere #self.lrus. why not self.lrus[#self.lrus] = lru here?

I find insert more concise here, and `#self.lrus` more concise in a for loop... That's all 😅 

> related: do we care enough about micro-optimizations to store an lrus table index as part of self, rather than than recomputing #lrus

Thought about it but, hardly... Those are not very hot code paths, only `:get()` should be.

**Comment by thibaultcha** [comment]
@Tieske anything to add here?

**Comment by Tieske** [comment]
I do not get this feature? What am I missing?

If you want a different lru, then you'd call `new` again, and you'd get a different lru. So why one object with multiple lru's? unnecessary complex? using a ffi based one would be one extra option at create time.

If anything, a Lua land cache should be "named", and any key stored in the lru cache should be stored in the shm as `lru_cache_name .. ":" .. key` to prevent collisions when using multiple lru caches on top of a single shm.

**Comment by Tieske** [comment]
and additional to that, it should use a named ipc channel as well. Such that an lru cache would only receive a callback for its own cache, not for other ones.

**Comment by thibaultcha** [comment]
> I do not get this feature? What am I missing?

You suggested it at #8, and I thought it was a good idea. I still do.

> If you want a different lru, then you'd call new again, and you'd get a different lru. So why one object with multiple lru's? unnecessary complex? using a ffi based one would be one extra option at create time.

That is not how you originally suggested it, and I do prefer the current approach, which is lighter: a single mlcache means only 1 IPC, less channels, less overhead, and more flexibility than just an option for the FFI implementation.

**Comment by thibaultcha** [comment]
@Tieske I have no idea what your other comments are about, since they are not on the files lines themselves but in the commit.

**Comment by thibaultcha** [comment]
I will assume where they were intended to be.

> and additional to that, it should use a named ipc channel as well. Such that an lru cache would only receive a callback for its own cache, not for other ones.

No, this is by design since the `delete()` function does not accept the LRU name for user-friendliness. The cost is trivial.

> this whole library so far assumes that the data in Lua land cache and in shm are equal. There should be another callback between those two. The Lua land cache is not limited to scalar values and hence should be able to hold other data. But that requires a callback to convert the scalar shm based values to the lua-land presentation.

The Lua land cache indeed holds tables and scalar in their Lua form, I am not sure what this is referring to.

> This function is not symmetrical with shmlru_set below. The latter one will never return the cache-miss-sentinel value, while this one does.

Yes, this is by design.

**Comment by Tieske** [comment]
>>I do not get this feature? What am I missing?
>
>You suggested it at #8, and I thought it was a good idea. I still do.
>
>>If you want a different lru, then you'd call new again, and you'd get a different lru. So why one object with multiple lru's? unnecessary complex? using a ffi based one would be one extra option at create time.
>
>That is not how you originally suggested it, and I do prefer the current approach, which is lighter: a single mlcache means only 1 IPC, less channels, less overhead, and more flexibility than just an option for the FFI implementation.

Then I probably expressed myself poorly. I consider the mlcache-objects just a wrapper around an lru cache, with automatic sync added to shm and other workers. So probably I mixed mlcache-objects and lru in that message, as I see them in a 1-on-1 relation.

Because you'd want individual lru caches, but shared shm's, I'd expect something like:
```lua
local consumer_cache = mlcache.new("dbcache", { ipc_shm = "ipc_events" })
local api_cache      = mlcache.new("dbcache", { ipc_shm = "ipc_events" })
local plugin_cache   = mlcache.new("dbcache", { ipc_shm = "ipc_events" })
```
If using this with a 16 worker system, each event will invoke 48 callbacks, where only 16 would be necessary. That is because the channel is just one generic channel, and not lru-cache specific.

So by naming the individual lru-caches, and then assigning individual channels a lot of callback calls will be saved. Additionally, by namespacing in the shm based on the you'll avoid key collisions in the shm (which is an important user expectation criteria imo)

What usecase do you see for the current implementation, with "nested lru's"? eg:
```lua
local something = consumer_cache:get(
    key, 
    { lru = "another_lru_name" },
    function()
        -- some callback implementation
    end)
```
This seems rather cumbersome to me. But apparently I haven't figured out your design concept just yet :smile:.

**Comment by Tieske** [comment]
>>this whole library so far assumes that the data in Lua land cache and in shm are equal. There should be another callback between those two. The Lua land cache is not limited to scalar values and hence should be able to hold other data. But that requires a callback to convert the scalar shm based values to the lua-land presentation.
>
>The Lua land cache indeed holds tables and scalar in their Lua form, I am not sure what this is referring to.

simple example, assume the callback for fetching data from the database is returning an array as a lua table:
```lua
    data = { "birds", "cows", "Flying Elephants" }
```
This data is serialised and stored in the shm. Now I want a second callback between the shm and the lru-cache like this:
```lua
-- implement hash-part for case-insensitive lookup
local function shm_to_lua_cb(data)
    if not data then
        return data
    end
    for i, v in ipairs(data) do
        data[v:lower()] = i
    end
    return data
end
```
The result being:
```lua
    -- shm entry
    data = {
        [1] = "birds", 
        [2] = "cows",
        [3] = "Flying Elephants",
    }

    -- lru entry
    data = {
        [1] = "birds", 
        [2] = "cows",
        [3] = "Flying Elephants",
        ["birds"] = 1,
        ["cows"] = 2,
        ["flying elephants"] = 3,
    }
```
This is a very simple reverse lookup example, but it could do more complex stuff like deserialising into objects, etc.

**Comment by thibaultcha** [comment]
#16 turns mlcache into an LRU extension backed up by an shm that can be shared between mlcache instances :)
